File: data_structures/LinkedList/python/doubly_linked_list.py
from linkedlist import Node, LinkedList

class DoublyNode(Node):
    def __init__(self, value):
        super().__init__(value)
        self._prev = None
    def getPrev(self):
        return self._prev
    def setPrev(self, value):
        self.__checkIsValidNode(value)
        self._prev = value

class DoublyLinkedList(LinkedList):
    _top: DoublyNode | None
    _last: DoublyNode | None

    def __init__(self):
        super().__init__()
        self._last = None
    
    """
        TODO: Add a reference to the last element in the list to traverse it in reverse order
    """
    def add(self, value):
        newNode = DoublyNode(value)
        if (isinstance(self._top, DoublyNode)):
            self._top.setPrev(newNode)
        if (self._last == None):
            self._last = self._top
        newNode.setNext(self._top)
        self._top = newNode
        self._size += 1

        # if (self._last == None):
        #     self._last = newNode
        # else:
        #     self._last.setPrev()
            

File: data_structures/LinkedList/typescript/doubly_linked_list.ts
import { LinkedList, Node_ } from "./linkedlist"

class DoublyNode<T> extends Node_<T> {
    private previous: null | DoublyNode<T>;
    constructor(value: T) {
        super(value);
        this.previous = null;
    }
    
    public setPrevious(previousNode: DoublyNode<T>) {
        this.previous = previousNode;
    }

}

class DoublyLinkedList<T> extends LinkedList<T> {
    // protected head: null | DoublyNode<T>;
    // protected tail: null | DoublyNode<T>;

    constructor() {
        super()
        // this.head = this.tail = null;
    }

    public add(nodeValue: T): void {
        const newNode = new DoublyNode<T>(nodeValue);
        if (this.isEmpty()) {
            this.head = this.tail = newNode;
        } else {
            this.tail!.setNext(newNode);
            newNode.setPrevious(this.tail as DoublyNode<T>);
            this.tail = newNode;
        }
    }

    public getValues(reverse?: 'reverse'): void {
        if (!reverse) {
            super.getValues()
        } else {
            
        }

    }
}

const doublyLL = new DoublyLinkedList();
doublyLL.add(1)
doublyLL.add(2)
doublyLL.add(3)
doublyLL.add(4)
doublyLL.getValues()